internal/providers/registry.go

```diff
@@ -1,52 +1,318 @@
+	r.actions <- func() {
+		// TODO: consider adding a map of fcAddr->providerID
+		for _, info := range r.providers {
+			if info.FcAddress == fcAddr {
+				infoChan <- info
+				break
+			}
 		}
-		return true
+		close(infoChan)
 	}
-	for i := range info.Addresses {
-		if addr == info.Addresses[i] {
-			return false
+
+	return <-infoChan
+}
+
+func (r *Registry) ProviderInfo(providerID peer.ID) *ProviderInfo {
+	infoChan := make(chan *ProviderInfo)
+	r.actions <- func() {
+		info, ok := r.providers[providerID]
+		if ok {
+			infoChan <- info
 		}
+		close(infoChan)
 	}
-	info.Addresses = append(info.Addresses, addr)
-	return true
+
+	return <-infoChan
+}
+
+func (r *Registry) AllProviderInfo() []*ProviderInfo {
+	var infos []*ProviderInfo
+	done := make(chan struct{})
+	r.actions <- func() {
+		infos = make([]*ProviderInfo, len(r.providers))
+		var i int
+		for _, info := range r.providers {
+			infos[i] = info
+			i++
+		}
+		close(done)
+	}
+	<-done
+	return infos
+}
+
+func (r *Registry) syncStartDiscover(fcAddr string, signature, signed []byte, errCh chan<- error) {
+	err := r.syncNeedDiscover(fcAddr)
+	if err != nil {
+		r.syncEndDiscover(fcAddr, nil, err, errCh)
+		return
+	}
+
+	// Mark discovery as in progress
+	r.discoTimes[fcAddr] = time.Time{}
+
+	r.pendingDisco++
+
+	// Do discovery asynchronously; do not block other discovery requests
+	go func() {
+		discoData, discoErr := r.discover(fcAddr, signed, signature)
+		r.actions <- func() {
+			r.syncEndDiscover(fcAddr, discoData, discoErr, errCh)
+			r.pendingDisco--
+		}
+	}()
+}
+
+func (r *Registry) syncEndDiscover(fcAddr string, discoData *discovery.Discovered, err error, errCh chan<- error) {
+	r.discoTimes[fcAddr] = time.Now()
+
+	if err != nil {
+		log.Errorw("could not discover provider", "err", err)
+		errCh <- err
+		close(errCh)
+		return
+	}
+
+	info := &ProviderInfo{
+		AddrInfo:  discoData.AddrInfo,
+		FcAddress: fcAddr,
+	}
+
+	r.syncRegister(info, errCh)
+}
+
+func (r *Registry) syncRegister(info *ProviderInfo, errCh chan<- error) {
+	r.providers[info.AddrInfo.ID] = info
+	err := r.syncPersistProvider(info)
+	if err != nil {
+		log.Errorw("could not persiste provider", "err", err)
+		errCh <- err
+	}
+	close(errCh)
+}
+
+func (r *Registry) syncPersistProvider(info *ProviderInfo) error {
+	if r.dstore == nil {
+		return nil
+	}
+	value, err := json.Marshal(info)
+	if err != nil {
+		return err
+	}
+	dsKey := datastore.NewKey(info.AddrInfo.ID.String())
+
+	return r.dstore.Put(dsKey, value)
+}
+
+func (r *Registry) syncNeedDiscover(fcAddr string) error {
+	completed, ok := r.discoTimes[fcAddr]
+	if ok {
+		// Check if discovery already in progress
+		if completed.IsZero() {
+			return ErrInProgress
+		}
+
+		// Check if last discovery completed too recently
+		if !r.policy.CanRediscover(completed) {
+			return ErrTooSoon
+		}
+	}
+	return nil
+}
+
+func (r *Registry) discover(fcAddr string, signature, signed []byte) (*discovery.Discovered, error) {
```

Comment:
the mix of `fcAddr` vs peer addresses, where it's unclear how much of this is filecoin specific, and what is generic to all providers should probably get clarified / delineated better

Reply:
I need to use `fcAddr` (e.g. `t01000`) to discover the miner via lotus.  From that, I can get the provider's `peer.ID` and its list of `multiaddr`.  Maybe this should be a new type of `Multiaddr` where the protocol is `filecoin`, and I name this value `DiscoveryID`?

Reply:
Renamed `fcAddr` to `discoverAddr` and commented on the need for additional data to tell the indexer where/how to perform the discovery.